File: golang/maximum_units_on_a_truck.go
package main

import (
	"sort"
)

// Space : O(1)
// Time  : O(n log n)

func maximumUnits(boxTypes [][]int, truckSize int) int {
	ans := 0
	sort.SliceStable(boxTypes, func(i, j int) bool {
		return boxTypes[i][1] > boxTypes[j][1]
	})

	for i := 0; i < len(boxTypes); i++ {
		if truckSize-boxTypes[i][0] >= 0 {
			ans += boxTypes[i][0] * boxTypes[i][1]
			truckSize -= boxTypes[i][0]
		} else {
			ans += truckSize * boxTypes[i][1]
			truckSize = 0
		}
		if truckSize == 0 {
			break
		}
	}

	return ans
}

File: python3/maximum_units_on_a_truck.py
"""
Space : O(1)
Time  : O(n log n)
"""


class Solution:
    def maximumUnits(self, boxTypes: List[List[int]], truckSize: int) -> int:
        ans = 0
        cap = 0
        boxTypes.sort(key=lambda b: -b[1])
        
        for q, v in boxTypes:
            if q + cap <= truckSize:
                cap += q
                ans += (q * v)
            else:
                space = truckSize - cap
                cap += space
                ans += space * v
            if cap == truckSize:
                break 
        
        return ans
